src/merkle_tree/constraints.rs

```diff
@@ -1,96 +1,40 @@
+use crate::crh::TwoToOneCRHGadget;
+use crate::merkle_tree::Config;
+use crate::{CRHGadget, Path};
 use ark_ff::Field;
+use ark_r1cs_std::alloc::AllocVar;
+use ark_r1cs_std::boolean::Boolean;
+use ark_r1cs_std::eq::EqGadget;
+#[allow(unused)]
 use ark_r1cs_std::prelude::*;
+use ark_r1cs_std::ToBytesGadget;
 use ark_relations::r1cs::{Namespace, SynthesisError};
+use ark_std::borrow::Borrow;
 
-use crate::{
-    crh::{FixedLengthCRH, FixedLengthCRHGadget},
-    merkle_tree::*,
-};
-
-use core::borrow::Borrow;
-
-pub struct PathVar<P, HGadget, ConstraintF>
+/// Represents a merkle tree path gadget.
+pub struct PathVar<P, LeafH, TwoToOneH, ConstraintF>
 where
     P: Config,
-    HGadget: FixedLengthCRHGadget<P::H, ConstraintF>,
+    LeafH: CRHGadget<P::LeafHash, ConstraintF>,
+    TwoToOneH: TwoToOneCRHGadget<P::TwoToOneHash, ConstraintF>,
     ConstraintF: Field,
 {
-    path: Vec<(HGadget::OutputVar, HGadget::OutputVar)>,
+    /// `path[i]` is 0 (false) iff ith non-leaf node from top to bottom is left.
+    path: Vec<Boolean<ConstraintF>>,
+    /// `auth_path[i]` is the entry of sibling of ith non-leaf node from top to bottom.
+    auth_path: Vec<TwoToOneH::OutputVar>,
+    /// THe sibling of leaf.
+    leaf_sibling: LeafH::OutputVar,
+    /// position of leaf. Should be 0 (false) iff leaf is on the left.
```

Comment:
```suggestion
    /// Which child is this leaf?.
```

Reply:
I would also rename the field to `is_leaf_right_child`

Reply:
fixed